include/simdjson/dom/document_stream.h

```diff
@@ -93,7 +94,17 @@ class document_stream {
      * @param other the end iterator to compare to.
      */
     really_inline bool operator!=(const iterator &other) const noexcept;
-
+    /**
+     * Gives the current index in the input document in bytes.
+     *
+     *   auto stream = parser.parse_many(json,window);
+     *   auto i = stream.begin();
+     *   for(; i != stream.end(); ++i) {
+     *      auto doc = *i;
+     *      size_t index = i.current_index();
+     *   }
+     */
+     really_inline size_t current_index() noexcept;
```

Comment:
Why put current_index in the iterator? Because creating the iterator "starts" the parser... before that point, the state of the document_stream does not support very well the notion of an index.

Reply:
This is going to be confusing to users, who don't normally ever see the iterator exposed. And it removes their ability to use C++ iteration effectively.

The simplest thing to do is to add this method to the document_stream. It would let users use iteration, but they wouldn't be able to do for (auto x : parser.parse_many()) because they would need the document_stream to be in a variable.

I think the right thing here is a new API with a new iterator, because we need the iterator to yield more information. We can't extend element because when the simdjson_result<element> is an error, the user still needs the index, correct?

So we do this instead:

The goal would be this API:

```c++
for (auto [elem, error, index] : parser.parse_many(json).with_indexes()) {
}
```

It means creating a new tiny class to return from with_indexes() whose only point is to have a begin() and end() that return this class:

```c++
class with_indexes_result : public simdjson_result<element> {
public:
  really_inline with_indexes_result(element&& elem,size_t start_loc) noexcept : simdjson_result<element>(std::forward<dom::element>(elem)), _start_index{start_index} {}
  really_inline with_indexes_result(error_code error,size_t start_index) noexcept : simdjson_result<element>(error), _start_index{start_index} {}
  really_inline size_t start_index() const noexcept { return _start_index; }
private:
  size_t _start_index;
};
```

Reply:
@lemire thinking through this more, the API you made inside the iterator is something I could probably live with becoming public, even if I'd rather have an iterator-friendly API in the future.

Reply:
I'll mark it private and indicate that it is subject to change.